### DISCUSSION TITLE
Lock down node version to exact version e.g. 16.16.0

### DISCUSSION THREAD

**Discussion by mcmxcdev** [vercel/vercel#8238] [Help] [answered]
In Netlify, it is possible to set the exact node version like described here: https://docs.netlify.com/configure-builds/manage-dependencies/#node-js-and-javascript

I couldn't find any documentation on how to do this in Vercel?

Especially for a CI run, I don't like using a loose Node version range like `16.x` which can be selected from the Vercel UI.

**Comment by amyegan** [comment]
_This was originally marked as the answer but is incorrect. Please see https://github.com/vercel/vercel/discussions/8238#discussioncomment-3710163_

---

Hi @mcmxcdev. The UI only allows selection of major versions of Node because Vercel automatically pulls in the latest minor and patch updates. But you are able to set a specific version in `package.json` by adding the [`engines.node` property](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#engines). For example, 
```
"engines": {
  "node": "16.16.0"
},
```

More details about Node version selection at https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-version

**Reply by IGassmann** [reply]
@amyegan this doesn't seem to work anymore. I get the following error with this config:

```
Warning: Detected "engines": { "node": "16.16.0" } in your `package.json` with major.minor.patch, but only major Node.js Version can be selected. Learn More: http://vercel.link/node-version
```

**Reply by IGassmann** [reply]
@amyegan How can we ensure we're using the same Node.js exact version locally as the one used on Vercel?

**Reply by amyegan** [reply]
Hi @IGassmann. My example was incorrect 😞

> Defining the node property inside [engines](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#engines) of a package.json file will override the selection made in the Project Settings and print a Build Step warning if the version does not match. Please avoid greater than ranges, such as >14.x or >=14, because this will match semver major versions of Node.js once available which contains breaking changes.
– [Node.js Runtime doc](https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-version)

**Comment by amyegan** [comment, marked as answer]
Only major versions can be specified. Since some minor and patch versions may have security or reliability issues, Vercel uses newer minor and patch updates when needed.

>Only major versions are available. That's because Vercel will automatically roll out minor and patch updates if needed (for example in the case that a security issue needs to be fixed).
>
> Defining the node property inside [engines](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#engines) of a package.json file will override the selection made in the Project Settings and print a Build Step warning if the version does not match.

– [Node.js Runtime documentation](https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-version)

**Comment by blentz100** [comment]
@amyegan Thanks. I was getting this warning during building:

```
Running "vercel build"
--
11:34:56.748 | Vercel CLI 33.1.0
11:34:57.174 | Warning: Detected "engines": { "node": ">=16.x" } in your `package.json` that will automatically upgrade when a new major Node.js Version is released. Learn More: http://vercel.link/node-version
11:34:57.181 | Installing dependencies...
11:34:57.449 | yarn install v1.22.17
11:34:57.489 | info No lockfile found.
11:34:57.501 | [1/5] Validating package.json...
11:34:57.505 | [2/5] Resolving packages...
```

and my package.json looked like this:

```
"engines": {
    "node": ">=16.x"
  },
```

Once I changed package.json to:

```
"engines": {
    "node": "18.x"
  },
```

it cleared the warning message at the build step. Hopefully this will help someone in the future.
